geoSegment.js

```diff
@@ -0,0 +1,353 @@
+const SESSION_SERIES_DIRECTORY_PATH = './sessionseries';
+const OUTPUT_DIRECTORY_PATH = './output';
+
+// Parameters controlling retry backoff
+const RETRY_BACKOFF_MIN = 1;
+const RETRY_BACKOFF_MAX = 1024; // ~ 17 mins
+const RETRY_BACKOFF_EXPONENTIATION_RATE = 2;
+const MIN_REQUEST_DELAY_SECONDS = 0.1;
+
+const SESSION_SERIES_FILE_PATHS = [];
+
+// Event Schedule generation
+const SCHEDULE_TYPE = 'Schedule';
+const MINIMAL_EVENT_SCHEDULE_SCHEMA = Joi.object().keys({
+  // If this array has length 0, an infinite loop will occur in the date
+  // generation algorithm below.
+  byDay: Joi.array().items(Joi.string()).min(1).required(),
+  startTime: Joi.string().required(),
+  endTime: Joi.string().required(),
+  idTemplate: Joi.string().required(),
+});
+
+/**
+ * @typedef {import('axios').AxiosError} AxiosError
+ * @typedef {import('axios').AxiosRequestConfig} AxiosRequestConfig
+ * @typedef {import('axios').AxiosResponse} AxiosResponse
+ * @typedef {{
+ * location?: {
+ *  geo?: {
+ *   latitude: number,
+ *   longitude: number,
+ *   }
+ *  },
+ * [k:string]: unknown
+ * }} SessionSeriesData
+ * @typedef {{
+ *  id: string,
+ *  state: 'updated' | 'deleted',
+ *  kind: 'SessionSeries',
+ *  data: SessionSeriesData
+ * }} SessionSeriesItem
+ * @typedef {{
+ *  startDate?: string,
+ *  superEvent?: string,
+ *  [k:string]: unknown
+ * }} ScheduledSessionData
+ * @typedef {{
+ *  id: string,
+ *  state: 'updated' | 'deleted',
+ *  kind: 'ScheduledSession',
+ *  data: ScheduledSessionData
+ * }} ScheduledSessionItem
+ * @typedef {{
+ *  latitude: number,
+ *  longitude: number,
+ *  radius: number,
+ *  identifier: string
+ * }} Segment
+ */
+
+/**
+ * @param {Segment[]} segments
+ */
+async function createSegmentDirectories(segments) {
+  // Create segment directories;
+  segments.map((segment) => segment.identifier)
+    .map(async (segmentIdentifier) => {
+      const segmentDirectoryPath = `${OUTPUT_DIRECTORY_PATH}/${segmentIdentifier}`;
+      await emptyOrMakeDirectory(segmentDirectoryPath);
+
+      // Create index file
+      await fsPromises.writeFile(`${OUTPUT_DIRECTORY_PATH}/${segmentIdentifier}/index.txt`, '');
+    });
+}
+
+/**
+ * @param {ScheduledSessionData} scheduledSessionData
+ * @param {SessionSeriesData} sessionSeries
+ */
+async function linkScheduledSessionAndSessionSeriesAndWrite(scheduledSessionData, sessionSeries) {
+  // Link ScheduledSession with SessionSeries
+  const linkedScheduledSessionData = {
+    ...scheduledSessionData,
+    superEvent: sessionSeries,
+  };
+
+  // Write ScheduledSession into each output segment directory
+  for (const segmentIdentifier of sessionSeries['imin:segment']) {
+    const filename = `${encodeURIComponent(linkedScheduledSessionData.id)}.json`;
+    await fs.writeJson(`${OUTPUT_DIRECTORY_PATH}/${segmentIdentifier}/${filename}`, linkedScheduledSessionData);
+    await fs.appendFile(`${OUTPUT_DIRECTORY_PATH}/${segmentIdentifier}/index.txt`, `${filename}\r\n`);
```

Comment:
Is there a bit logic missing here to only write to index.txt if the file does not already exist (i.e. it has not been created previously from the Schedule)? So we don't end up with duplicate entries in index.txt?

Reply:
Good catch, done